terraform/state.go

```diff
@@ -801,6 +802,32 @@ func (s *BackendState) Empty() bool {
 	return s == nil || s.Type == ""
 }
 
+// Rehash returns a unique content hash for this backend's configuration
+// as a uint64 value.
+// The Hash stored in the backend state needs to match the config itself, but
+// we need to compare the backend config after it has been combined with all
+// options.
+// This function must match the implementation used by config.Backend.
+func (s *BackendState) Rehash() uint64 {
+	if s == nil {
+		return 0
+	}
+
+	// Use hashstructure to hash only our type with the config.
+	code, err := hashstructure.Hash(map[string]interface{}{
+		"type":   s.Type,
+		"config": s.Config,
+	}, nil)
+
+	// This should never happen since we have just some basic primitives
+	// so panic if there is an error.
+	if err != nil {
+		panic(err)
+	}
+
+	return code
```

Comment:
I'd prefer here instead of copy/pasting the config code to actually generate a `config.Backend` and call Rehash. That way we only need to ensure that Rehash is bug free in one place. Although, if the hash ever changes its a big problem. :) But at least it limits that to one code path.